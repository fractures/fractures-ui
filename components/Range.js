import { Component, Fragment } from "react"

const defaultProps = {
	set: () => null,
	value: 0
}

class Range extends Component {
	render() {
		return (
			<Fragment>
				<input
					type="range"
					className="fr-range w-100"
					defaultValue={ this.props.value }
					onChange={ e => this.props.set(e.target.value) }
				/>
				<style jsx global>{`
					.fr-range {
						-webkit-appearance: none;

						background-color: transparent;
					}

					.fr-range:focus {
						outline: none;
					}

					.fr-range::-webkit-slider-runnable-track {
						height: 0.5rem;
						width: 100%;

						background-color: var(--shade-1);
						border-radius: 1rem;

						cursor: pointer;
					}

					.fr-range::-webkit-slider-thumb {
						height: 1.125rem;
						width: 1.125rem;

						border-radius: 1rem;
						border: 0;

						background: var(--shade-4);
						cursor: pointer;
						transform: translateY(-0.375rem);
						-webkit-appearance: none;
					}

					.fr-range:focus::-webkit-slider-thumb {
						box-shadow: 0 0 0 0.15rem var(--shade-2);
					}
				`}</style>
			</Fragment>
		)
	}
}

Range.defaultProps = defaultProps

export default Range
