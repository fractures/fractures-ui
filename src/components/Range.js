import React, { Component, Fragment } from "react"

const defaultProps = {
	set: () => null,
	value: 0
}

class Range extends Component {
	render() {
		return (
			<Fragment>
				<input
					type="range"
					className="fr-range w-100"
					defaultValue={ this.props.value }
					onChange={ e => this.props.set(e.target.value) }
				/>
				<style jsx global>{`
					.fr-range {
						-webkit-appearance: none;

						background-color: transparent;
					}

					.fr-range:focus {
						outline: none;
					}

					.fr-range::-webkit-slider-runnable-track {
						height: 0.5rem;
						width: 100%;

						background-color: var(--fr-100);
						border-radius: 1rem;

						cursor: pointer;
					}

					.fr-range::-webkit-slider-thumb {
						height: 1.125rem;
						width: 1.125rem;

						border-radius: 1rem;
						border: 0;

						background: var(--fr-500);
						cursor: pointer;
						transform: translateY(-0.375rem);
						-webkit-appearance: none;
					}

					.fr-range:focus::-webkit-slider-thumb {
						box-shadow: 0 0 0 0.1875rem var(--fr-focus);
					}
				`}</style>
			</Fragment>
		)
	}
}

Range.defaultProps = defaultProps

export default Range
