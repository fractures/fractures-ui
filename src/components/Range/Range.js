import classNames from "@sindresorhus/class-names"
import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"

const propTypes = {
	className: PropTypes.string,
	isVerbose: PropTypes.bool,
	max: PropTypes.number.isRequired,
	min: PropTypes.number.isRequired,
	onChange: PropTypes.func.isRequired,
	step: PropTypes.number,
	value: PropTypes.number
}

const defaultProps = {
	isVerbose: false,
	max: 100,
	min: 0,
	onChange: () => null,
	step: null,
	value: 0
}

const NakedRange = props => {
	const nakedClasses = classNames("hl-range", {
		"hl-range--verbose": props.isVerbose
	})

	return (
		<input
			className={ `${ props.className } ${ nakedClasses }` }
			defaultValue={ props.value }
			onChange={ e => props.onChange(e) }
			type="range"
			step={ props.step }
			min={ props.min }
			max={ props.max }
		/>
	)
}

const NakedRangeWrap = styled(NakedRange)`
	position: relative;

	display: flex;
	margin: 0;
	padding: 0;
	width: 100%;

	&.hl-range--verbose:before,
	&.hl-range--verbose:after {
		display: block;
		height: 1.5rem;
		padding: 0 0.5rem;

		background-color: var(--color-gray-100);
		border-radius: 1rem;
		color: var(--color-gray-500);

		font-size: 0.75rem;
		line-height: 1.5rem;
		text-align: center;
	}

	&.hl-range--verbose:after {
		right: 0;

		margin-left: 0.5rem;

		content: attr(max);
	}

	&.hl-range--verbose:before {
		left: 0;

		margin-right: 0.5rem;

		content: attr(min);
	}

	&:focus {
		outline: none;
	}

	&::-webkit-slider-runnable-track {
		height: 0.5rem;
		width: 100%;

		background-color: var(--color-gray-100);
		border-radius: 1rem;

		cursor: pointer;
	}

	&::-webkit-slider-thumb {
		position: relative;
		z-index: 1;

		height: 1rem;
		width: 1rem;
		-webkit-appearance: none;

		border-radius: 1rem;
		border: 0;

		background: var(--color-gray-500);
		cursor: pointer;
		transform: translateY(calc(-0.5rem + 0.25rem));
	}

	&:focus::-webkit-slider-thumb {
		box-shadow: 0 0 0 0.1875rem var(--color-gray-100);
	}
`

const NakedRangeContainer = props => (
	<div className={ props.className }>
		<NakedRangeWrap { ...props } />
		{props.isVerbose && <div className="hl-range__value">{props.value}</div>}
	</div>
)

const Range = styled(NakedRangeContainer)`
	position: relative;

	display: flex;
	width: 100%;

	background-color: transparent;
	-webkit-appearance: none;

	.hl-range__value {
		display: block;
		height: 1.5rem;
		padding: 0 0.5rem;
		margin-left: 0.5rem;

		background-color: var(--color-gray-500);
		border-radius: 1rem;
		color: var(--color-gray-100);

		font-size: 0.75rem;
		font-weight: bold;
		line-height: 1.5rem;
		text-align: center;
	}
`

NakedRange.propTypes = propTypes
NakedRange.defaultProps = defaultProps

Range.propTypes = propTypes
Range.defaultProps = defaultProps

export default Range
