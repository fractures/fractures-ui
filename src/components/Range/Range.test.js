import { Range } from "./Range.js"

const mockFn = jest.fn()

describe("Range", () => {
	const componentRender = <Range value={ 42 } set={ mockFn() } />
	const component = renderer.create(componentRender)
	const tree = component.toJSON()

	it("Renders correctly", () => {
		expect(tree).toMatchSnapshot()
	})
})
