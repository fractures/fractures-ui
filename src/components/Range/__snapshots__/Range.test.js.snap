// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range Renders correctly 1`] = `
<div
  className="sc-bwzfXH AHMGE"
>
  <input
    className="sc-bwzfXH AHMGE sc-bdVaJa hNhBIV hl-range"
    defaultValue={42}
    max={100}
    min={0}
    onChange={[Function]}
    step={null}
    type="range"
  />
</div>
`;
